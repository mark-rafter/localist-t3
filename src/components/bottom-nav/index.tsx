import { BottomNavLink } from "@/components/bottom-nav/bottom-nav-link";
import { useSession } from "next-auth/react";
import {
  HiHeart,
  HiHome,
  HiInboxArrowDown,
  HiOutlineHeart,
  HiOutlineHome,
  HiOutlineInboxArrowDown,
  HiOutlinePencilSquare,
  HiOutlineUser,
  HiPencilSquare,
  HiUser,
} from "react-icons/hi2";

export function BottomNav() {
  const { status: sessionStatus } = useSession();
  if (sessionStatus !== "authenticated") {
    return <></>;
  }
  return (
    <div className="fixed bottom-4 left-1/2 z-50 h-16 w-full max-w-lg -translate-x-1/2 rounded-full border border-gray-600 bg-gray-800 sm:hidden">
      <div className="grid h-full grid-cols-5">
        <BottomNavLink
          pos="start"
          icon={HiOutlineHome}
          filledIcon={HiHome}
          label="Feed"
        />
        <BottomNavLink
          icon={HiOutlineInboxArrowDown}
          filledIcon={HiInboxArrowDown}
          label="Messages"
          badge="3"
        />
        <BottomNavLink
          icon={HiOutlinePencilSquare}
          filledIcon={HiPencilSquare}
          label="New Post"
        />
        <BottomNavLink
          icon={HiOutlineHeart}
          filledIcon={HiHeart}
          label="Likes"
        />
        <BottomNavLink
          pos="end"
          icon={HiOutlineUser}
          filledIcon={HiUser}
          label="Profile"
        />
      </div>
    </div>
  );
}
