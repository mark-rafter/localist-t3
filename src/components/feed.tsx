import Image from "next/image";
import React from "react";
import Link from "next/link";
import { HiOutlineClock, HiOutlineMapPin } from "react-icons/hi2";

const FeedItem = ({
  postId,
  distance,
  price,
  size,
}: {
  postId: number;
  distance: number;
  price: number;
  size: "XS" | "small" | "medium" | "large" | "XL" | "XXL";
}) => {
  const tailwindSmPixels = 384;
  return (
    <article className="max-w-sm overflow-hidden rounded-lg border border-gray-200 bg-white antialiased shadow dark:border-gray-700 dark:bg-gray-800">
      <Link href={`post/${postId}`} prefetch={false}>
        <Image
          className="rounded-t-lg"
          width={tailwindSmPixels}
          height={tailwindSmPixels}
          src="/tshirt-474px.jpg"
          alt=""
        />
      </Link>
      <div className="p-3 sm:p-4">
        <div className="flex items-baseline justify-between text-gray-700 dark:text-gray-300">
          <h3 className="text-sm uppercase tracking-wide">{size}</h3>
          <span className="inline-flex items-center rounded px-2 text-xs">
            <HiOutlineClock className="mr-1 h-3 w-3" />2 min.
          </span>
        </div>
        <Link href={`post/${postId}`} prefetch={false}>
          <h2 className="w-64 truncate pt-1 text-xl font-bold tracking-tight text-gray-900 dark:text-white">
            Unisex plain purple nylon tee, 3 years old
          </h2>
        </Link>
        <div className="flex items-center justify-between">
          <span className="text-xl font-bold text-gray-900 dark:text-white">
            Â£{price}
          </span>
          <span className="flex items-center rounded-lg text-sm font-normal text-gray-400">
            <span className="flex-1 whitespace-nowrap">{distance} miles</span>
            <HiOutlineMapPin className="h-4 w-4" />
          </span>
        </div>
      </div>
    </article>
  );
};

function getRandomInt(min: number, max: number) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min) + min); // The maximum is exclusive and the minimum is inclusive
}

export const Feed = () => (
  <div
    className="mb-4 grid grid-cols-1 justify-items-center gap-4 rounded-lg p-4 
    lg:grid-cols-2 xl:grid-cols-3 2xl:grid-cols-4"
  >
    {[...Array(8).keys()].map((k) => (
      <FeedItem
        key={k + 1}
        postId={k + 1}
        price={33}
        distance={22}
        size={"small"}
      />
    ))}
  </div>
);
