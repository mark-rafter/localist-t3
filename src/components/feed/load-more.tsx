import { Button, Spinner } from "flowbite-react";

export const LoadMore = ({
  disabled,
  onClick,
  loading,
}: {
  disabled: boolean;
  onClick: () => void;
  loading: boolean;
}) => (
  <div className="flex justify-center">
    <Button className="w-64" color="gray" onClick={onClick} disabled={disabled}>
      {loading ? (
        <>
          <Spinner size="sm" className="mr-3" light={true} />
          Loading...
        </>
      ) : (
        <>Load more</>
      )}
    </Button>
  </div>
);
