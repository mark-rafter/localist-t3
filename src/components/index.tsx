import { Layout } from "./layout";
import { LinkOrDisabled } from "./link-or-disabled";
import { SkeletonOrChildren } from "./skeleton-or-children";

export { LinkOrDisabled, SkeletonOrChildren, Layout };
