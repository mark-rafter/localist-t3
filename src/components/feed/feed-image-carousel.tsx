import { LinkOrDisabled } from "@/components";
import type { FeedItemProps } from "@/components/feed/feed-item";
import { Carousel } from "flowbite-react";
import Image from "next/image";
import React from "react";

type ImageCarouselProps = Pick<
  FeedItemProps,
  "id" | "images" | "title" | "isPreview"
>;

function ImagesContainer({
  imageLength,
  children,
}: React.PropsWithChildren<{ imageLength: number }>) {
  return imageLength == 1 ? (
    <>{children}</>
  ) : (
    <Carousel slideInterval={3000}>{children}</Carousel>
  );
}

export function FeedImageCarousel({
  id,
  title,
  images,
  isPreview,
}: ImageCarouselProps) {
  const smPixels = 384;

  if (!images) return <></>;

  return (
    <div>
      <ImagesContainer imageLength={images.length}>
        {images.map((image, index) => (
          <LinkOrDisabled
            key={index}
            target={`post/${id}`}
            disabled={isPreview == true}
            prefetch={false}
          >
            <Image
              src={image}
              alt={`${title} Image ${index + 1}`}
              width={smPixels}
              height={smPixels}
            />
          </LinkOrDisabled>
        ))}
      </ImagesContainer>
    </div>
  );
}
