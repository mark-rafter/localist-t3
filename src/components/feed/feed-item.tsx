import Image from "next/image";
import React from "react";
import { HiOutlineClock, HiOutlineMapPin } from "react-icons/hi2";
import { Carousel } from "flowbite-react";
import { LinkOrDisabled } from "@/components/link-or-disabled";

type DraftFeedItem = {
  title: string;
  size: "xs" | "small" | "medium" | "large" | "xl";
  brand: string;
  price: number;
  images?: string[];
};

type FeedItemProps = DraftFeedItem & {
  postId: number;
  distance: number;
  isPreview?: boolean;
};

export const FeedItem = ({
  postId,
  distance,
  title,
  price,
  size,
  images,
  isPreview,
}: FeedItemProps) => {
  const tailwindSmPixels = 384;
  return (
    <article className="max-w-sm overflow-hidden rounded-lg">
      {images && (
        <div className="h-56 bg-gray-500 sm:h-64 xl:h-80 2xl:h-96">
          <Carousel slideInterval={3000}>
            {[...Array(4).keys()].map((k) => (
              <LinkOrDisabled
                key={k}
                target={`post/${postId}`}
                disabled={isPreview == true}
              >
                <Image
                  src={`https://flowbite.com/docs/images/carousel/carousel-${
                    k + 1
                  }.svg`}
                  alt="..."
                  width={tailwindSmPixels}
                  height={tailwindSmPixels}
                />
              </LinkOrDisabled>
            ))}
          </Carousel>
        </div>
      )}

      <div className="bg-gray-800 p-3 antialiased sm:p-4">
        <div className="flex items-baseline justify-between text-gray-300">
          <h3 className="text-sm uppercase tracking-wide">{size}</h3>
          <span className="inline-flex items-center rounded px-2 text-xs">
            <HiOutlineClock className="mr-1 h-3 w-3" />2 min.
          </span>
        </div>
        <LinkOrDisabled target={`post/${postId}`} disabled={isPreview == true}>
          <h2 className="w-64 truncate pt-1 text-xl font-bold tracking-tight">
            {title}
          </h2>
        </LinkOrDisabled>
        <div className="flex items-center justify-between">
          <span className="text-xl font-bold">Â£{price}</span>
          <span className="flex items-center rounded-lg text-sm font-normal text-gray-400">
            <span className="flex-1 whitespace-nowrap">{distance} miles</span>
            <HiOutlineMapPin className="h-4 w-4" />
          </span>
        </div>
      </div>
    </article>
  );
};

export const DraftFeedItem = (props: DraftFeedItem) => {
  return <FeedItem postId={0} distance={0} {...props} />;
};
