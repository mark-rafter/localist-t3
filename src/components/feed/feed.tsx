import type { Coordinates } from "@/helpers/distance";
import { humanize } from "@/helpers/distance";
import { relativeTimeFromDates } from "@/helpers/relative-time";
import type { RouterOutputs } from "@/utils/api";
import { useSession } from "next-auth/react";
import { FeedItem, SkeletonFeedItem } from "./feed-item";
import { getCookieOrDefault } from "@/helpers/cookie";

export type FeedProps = {
  posts: RouterOutputs["post"]["getFeed"]["posts"];
  userCoords?: Coordinates;
};

export function SkeletonFeed() {
  return (
    <div
      className="mb-4 grid grid-cols-1 justify-items-center gap-4 rounded-lg p-4 lg:grid-cols-2 xl:grid-cols-3 2xl:grid-cols-4"
    >
      {[...Array(8).keys()].map((k) => (
        <SkeletonFeedItem key={k} />
      ))}
    </div>
  );
}

export function Feed({ posts, userCoords }: FeedProps) {
  return (
    <div
      className="mb-4 grid grid-cols-1 justify-items-center gap-4 rounded-lg p-4 lg:grid-cols-2 xl:grid-cols-3 2xl:grid-cols-4"
    >
      {posts.map((post) => {
        const postCoords = {
          lat: post.author.lat,
          long: post.author.long,
        };
        const distance = userCoords ? humanize(userCoords, postCoords) : "";
        const postAge = relativeTimeFromDates(post.createdAt);
        return (
          <FeedItem
            key={post.id}
            {...post}
            postAge={postAge}
            distance={distance}
          />
        );
      })}
    </div>
  );
}

export function ClientFeed({ posts }: FeedProps) {
  // todo: refactor into useUserCoords()
  const { data } = useSession();

  const userCoords = {
    lat: data?.user.lat ?? getCookieOrDefault("x-vercel-ip-latitude", 51.5),
    long: data?.user.long ?? getCookieOrDefault("x-vercel-ip-longitude", 0.0),
  };

  return <Feed posts={posts} userCoords={userCoords} />;
}
