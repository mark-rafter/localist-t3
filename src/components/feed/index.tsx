import type { Coordinates } from "@/helpers/distance";
import { humanize } from "@/helpers/distance";
import { relativeTimeFromDates } from "@/helpers/relative-time";
import { usePersistedState } from "@/hooks/use-persisted-state";
import type { RouterOutputs } from "@/utils/api";
import { useSession } from "next-auth/react";
import React from "react";
import { FeedItem } from "./feed-item";

export type FeedProps = {
  posts: RouterOutputs["post"]["getMany"]["posts"];
  cursor?: string;
};

export function Feed({ posts }: FeedProps) {
  const { data } = useSession();
  // todo: set geo cookie in middleware and extract defaults
  const [userCoords] = usePersistedState<Coordinates>("user-coords", {
    lat: data?.user.lat ?? 51.5,
    long: data?.user.long ?? 0.0,
  });

  return (
    <div
      className="mb-4 grid grid-cols-1 justify-items-center gap-4 rounded-lg p-4 
      lg:grid-cols-2 xl:grid-cols-3 2xl:grid-cols-4"
    >
      {posts.map((post) => {
        const postCoords = {
          lat: post.author.lat,
          long: post.author.long,
        };
        const distance = humanize(userCoords, postCoords);
        const postAge = relativeTimeFromDates(post.createdAt);
        return (
          <FeedItem
            key={post.id}
            {...post}
            postAge={postAge}
            distance={distance}
          />
        );
      })}
    </div>
  );
}
