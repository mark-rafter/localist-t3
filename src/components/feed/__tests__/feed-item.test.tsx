import { FeedItem } from "@/components/feed";
import { render } from "@testing-library/react";

it("renders HomePage unchanged", () => {
  const { container } = render(
    <FeedItem
      id={1}
      title="test"
      size="xs"
      price={123}
      images={[]}
      distance="X miles away"
      postAge="X days ago"
      isPreview={false}
    />
  );
  expect(container).toMatchSnapshot();
});
