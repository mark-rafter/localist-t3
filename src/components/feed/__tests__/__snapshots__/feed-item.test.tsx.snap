// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders HomePage unchanged 1`] = `
<div>
  <article
    class="max-w-sm overflow-hidden rounded-lg bg-gray-800"
  >
    <div>
      <div
        class="relative h-full w-full"
        data-testid="carousel"
      >
        <div
          class="flex h-full snap-mandatory overflow-y-hidden overflow-x-scroll scroll-smooth rounded-lg snap-x indiana-scroll-container indiana-scroll-container--hide-scrollbars"
        />
        <div
          class="absolute bottom-5 left-1/2 flex -translate-x-1/2 space-x-3"
        />
        <div
          class="absolute top-0 left-0 flex h-full items-center justify-center px-4 focus:outline-none"
        >
          <button
            class="group"
            data-testid="carousel-left-control"
            type="button"
          >
            <span
              class="inline-flex h-8 w-8 items-center justify-center rounded-full bg-white/30 group-hover:bg-white/50 group-focus:outline-none group-focus:ring-4 group-focus:ring-white dark:bg-gray-800/30 dark:group-hover:bg-gray-800/60 dark:group-focus:ring-gray-800/70 sm:h-10 sm:w-10"
            >
              <svg
                class="h-5 w-5 text-white dark:text-gray-800 sm:h-6 sm:w-6"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15 19l-7-7 7-7"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="absolute top-0 right-0 flex h-full items-center justify-center px-4 focus:outline-none"
        >
          <button
            class="group"
            data-testid="carousel-right-control"
            type="button"
          >
            <span
              class="inline-flex h-8 w-8 items-center justify-center rounded-full bg-white/30 group-hover:bg-white/50 group-focus:outline-none group-focus:ring-4 group-focus:ring-white dark:bg-gray-800/30 dark:group-hover:bg-gray-800/60 dark:group-focus:ring-gray-800/70 sm:h-10 sm:w-10"
            >
              <svg
                class="h-5 w-5 text-white dark:text-gray-800 sm:h-6 sm:w-6"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9 5l7 7-7 7"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="p-3 antialiased sm:p-4"
    >
      <div
        class="flex items-baseline justify-between text-gray-300"
      >
        <h3
          class="text-sm uppercase tracking-wide"
        >
          xs
        </h3>
        <span
          class="inline-flex items-center px-1 text-xs"
        >
          <span
            class="mr-1 flex-1 whitespace-nowrap"
          >
            X days ago
          </span>
          <svg
            aria-hidden="true"
            class="h-3 w-3"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 6v6h4.5m4.5 0a9 9 0 11-18 0 9 9 0 0118 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
      </div>
      <a
        href="post/1"
      >
        <h2
          class="w-64 truncate pt-1 text-xl font-bold tracking-tight"
        >
          test
        </h2>
      </a>
      <div
        class="flex items-center justify-between"
      >
        <span
          class="text-xl font-bold"
        >
          Â£
          123
        </span>
        <span
          class="flex items-center text-sm font-normal text-gray-400"
        >
          <span
            class="mr-1 flex-1 whitespace-nowrap"
          >
            X miles away
          </span>
          <svg
            aria-hidden="true"
            class="h-4 w-4"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 10.5a3 3 0 11-6 0 3 3 0 016 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M19.5 10.5c0 7.142-7.5 11.25-7.5 11.25S4.5 17.642 4.5 10.5a7.5 7.5 0 1115 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
      </div>
    </div>
  </article>
</div>
`;
